[CIR] Add option to emit MLIR in LLVM dialect.

Jezurko opened the pull request:
This PR adds the flag `-emit-mlir-llvm` to allow emitting of MLIR in the LLVM dialect (cc @xlauko who asked me to do this).
I'm not sure if the naming of the flag is the best and maybe someone will have a better idea.
Another solution would be to make the `-emit-mlir` flag have a value, that specifies the target dialect (CIR/MLIR std dialects/LLVM Dialect).

xlauko commented:
Instead of the current approach, I recommend using `-emit-mlir=std` to match the behavior of `-emit-mlir`, which is ambiguous regarding which MLIR dialect is being referenced. And `-emit-mlir=llvm` should be used to target the LLVM dialect.

For the case where `-emit-mlir=llvm -fclangir-direct-lowering` is used, it should directly lower from CIR to LLVM, with `-fclangir-direct-lowering` remaining an ON by default. And `-emit-mlir=llvm -fno-clangir-direct-lowering` would proceed through the standard dialects.

bcardosolopes commented:
I like @xlauko idea, couple things to add to the discussion:
- `emit-mlir` should be acceptable without additional arguments and default to std for no-direct and llvm for direct.
- We need driver tests as part of the PR (check the produced cc1 invocation).
- We need a warning if `-emit-mlir=std` is used with `-fclangir-direct-lowering`

bcardosolopes commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3120,7 +3120,16 @@ def emit_cir_only : Flag<["-"], "emit-cir-only">,
   Group<Action_Group>, HelpText<"Similar to -emit-cir but also lowers structured CFG into basic blocks.">;
 def emit_mlir : Flag<["-"], "emit-mlir">, Visibility<[CC1Option]>, Group<Action_Group>,
-  HelpText<"Build ASTs and then lower through ClangIR to MLIR, emit the .milr file">;
+  HelpText<"Build ASTs and then lower through ClangIR to MLIR (standard dialects "
```
Please indicate that the behavior depends on other flags (direct, through)

bcardosolopes commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3120,7 +3120,16 @@ def emit_cir_only : Flag<["-"], "emit-cir-only">,
+  HelpText<"Build ASTs and then lower through ClangIR to the selected MLIR dialect, emit the .mlir file. "
+  "Allowed values are `std` for MLIR standard dialects "
+  "`llvm` for the LLVM dialect and `cir` for the ClangIR dialect.">,
+  Values<"std,llvm,cir">,
```
It's confusing to have CIR here, better remove it. Should be std or LLVM (we already have `-emit-cir` and this PR is not unifying any of the existing handling either).

bcardosolopes commented on clang/lib/CIR/FrontendAction/CIRGenAction.cpp:
```diff
@@ -271,14 +270,45 @@ class CIRGenConsumer : public clang::ASTConsumer {
+        emitCIR();
+        break;
+      }
+      if (loweredMlirModule) {
```
What happens if this if is not taken?

bcardosolopes commented on clang/lib/CIR/FrontendAction/CIRGenAction.cpp:
```diff
@@ -271,14 +270,45 @@ class CIRGenConsumer : public clang::ASTConsumer {
-      flags.enableDebugInfo(/*enable=*/true, /*prettyForm=*/false);
-      loweredMlirModule->print(*outputStream, flags);
+      mlir::ModuleOp loweredMlirModule;
+      switch (feOptions.MLIRTargetDialect) {
```
After removing CIR support this can be simplified to not use a switch

bcardosolopes reviewed (changes requested):
Thanks for improving the PR, more comments inline. This still need tests (both driver test and a few simple IR output checks).

bcardosolopes commented:
To expand a bit on testing:
- Driver one: should check that `-emit-mlir` without and without `={std,mlir}` will expand to the right one passed down to `-cc1`. This can be done with `-###`.
- IR one: simple check that the outputed IR is the flavor we promise.

Jezurko commented:
> Driver one: should check that `-emit-mlir` without and without `={std,mlir}` will expand to the right one passed down to -`cc1`. This can be done with `-###`.

In the way I've done it `-emit-mlir` without value does not expand to the `-emit-mlir=` version. Instead, the value of the `MLIRTargetDialect` enum stays as `MLIR_Default` which signalizes that the action handler should decide based on the Direct/Indirect lowering flag (it has to check that flag to decide between direct/indirect lowering to llvm dialect anyway). Should I change that?

Jezurko commented:
I have updated the PR according to the comments.
I'm not entirely sure if the tests for the driver (via `-###`) are as envisioned. I'm happy to change them if required

bcardosolopes commented on clang/test/CIR/emit-mlir.c:
```diff
@@ -0,0 +1,23 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fno-clangir-direct-lowering -emit-mlir=llvm %s -o - | FileCheck %s -check-prefix=LLVM
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fno-clangir-direct-lowering -emit-mlir=std %s -o - | FileCheck %s -check-prefix=STD
+
+// RUN: %clang -fclangir -Xclang -emit-mlir %s -o - -### 2>&1 | FileCheck %s -check-prefix=OPTS_NO_VALUE
```
Thanks for the updates! Almost there, few more changes needed (sorry for not catching early):
- The version without the EQ should be driver only.
- cc1 only accepts the EQ version, requiring either llvm/std.
- We can ditch `MLIR_Default` by picking one of the EQ values at driver time.

andykaylor commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3120,7 +3120,16 @@ def emit_cir_only : Flag<["-"], "emit-cir-only">,
+def emit_mlir_EQ : Joined<["-"], "emit-mlir=">, Visibility<[CC1Option]>, Group<Action_Group>,
+  HelpText<"Build ASTs and then lower through ClangIR to the selected MLIR dialect, emit the .mlir file. "
+  "Allowed values are `std` for MLIR standard dialects and `llvm` for the LLVM dialect.">,
+  Values<"std,llvm">,
```
Can "cir" and "cir-flat" be added here?

bcardosolopes replied:
I asked to remove it but now in light of upstreaming discussion https://github.com/llvm/llvm-project/pull/127835#discussion_r1964324860, we should probably converge to make upstreaming approach easier. @Jezurko sorry for the noise. We should add `cir` and `cir-flat` to match was likely coming for Flang as part of that discussion. We should also rename `std` to `core`.

Andy, should we hold on this until you land to make your life easier or is it enough if @Jezurko applies the changes above?

andykaylor replied:
I think it makes sense to merge this first, then I can update my upstream patch to align with this and the wishes of the reviewers there.

andykaylor replied:
I have posted an update in https://github.com/llvm/llvm-project/pull/127835 that uses the `-emit-mlir=[cir|core]` option. I was referencing this PR when I implemented it, so it shouldn't be too far off. One difference is that I made the MLIR dialect a separate named enum.

bcardosolopes replied:
Added another small discussion point over there

Jezurko commented:
Sorry for the delay. I have modified the PR by adding the `cir` and `cir-flat` values to the flag and changed the cc1/driver availability of the flags, as requested.
One thing to consider is, if `-emit-cir` and `-emit-cir-flat` need to emit files ending with `*.cir`. If not,  I could make it so that the `-emit-cir` and `-emit-cir-flat` options are just aliases for the `-emit-mlir=` flag(and they would produce `*.mlir`).

I've noticed the code formatting failed check, but it's in a code that was already there before this PR. The suggested change seems rather ugly in my opinion, but I can add it to the PR.

bcardosolopes commented:
> One thing to consider is, if `-emit-cir` and `-emit-cir-flat` need to emit files ending with `*.cir`.

It should honor whatever is specified in `-o`. If `-o` isn't specified, it uses the `.o` extension by default (no CIR specific changes here, same for `-emit-llvm`). This is somewhat tricky right now because in theory we'd still want `.o` if we are emitting the serialized version, but we only emit the textual one at the moment. 

> If not, I could make it so that the `-emit-cir` and `-emit-cir-flat` options are just aliases for the `-emit-mlir=` flag (and they would produce `*.mlir` files).

I don't have a problem with that, but see considerations above.

> I've noticed the code formatting check failed, but it's in a code that was already there before this PR.

The code is already there but seems like you touched it? (at least I see a related change in Driver.cpp).

> The suggested change seems rather ugly in my opinion, but I can add it to the PR, if it is necessary to satisfy the formatter

Yea, it looks bad indeed, I'd just apply it and move on (perhaps it's a clang format bug?), let's see how it looks like once you update the PR.

Btw, seems like there are failing tests too.

andykaylor commented:
> If `-o` isn't specified, it uses the `.o` extension by default (no CIR specific changes here, same for `-emit-llvm`).

I don't think that's correct. `getOutputStream` in `clang/lib/CIR/FrontendAction/CIRGenAction.cpp` creates a filename for output based on the `CIRGenAction::OutputType` value passed in.

bcardosolopes commented:
> I don't think that's correct. `getOutputStream` in `clang/lib/CIR/FrontendAction/CIRGenAction.cpp` creates a filename for output based on the `CIRGenAction::OutputType` value passed in.

Yea, which in the example I tried gave me `.o`, the point being that it's not something that CIR is changing right now

bcardosolopes reviewed (approved):
Awesome, thank you!

bcardosolopes commented:
@andykaylor direction sounds good here regarding the work you did upstream? Just want to double check before I merge it

andykaylor reviewed (approved):
Sorry for the delay. I forgot this was waiting on me. Yes, I think this looks good.
